### ISSUE TITLE
The mode does not turn off

### ISSUE THREAD

**Issue by stasvlasov** [xahlee/xah-fly-keys#18]
When I switch the mode off with `xah-fly-keys-off` function it turns back on each time I use minibuffer. I suppose all hooks should be canceled within this function.

**Comment by xahlee** [comment]
sorry took so long. i tried to fix it for a hour but am not sure why it happens.

**Comment by stasvlasov** [comment]
Probably because of these lines (from line 3060):
``` emacs lisp
(add-hook 'minibuffer-setup-hook 'xah-fly-insert-mode-activate)
(add-hook 'minibuffer-exit-hook 'xah-fly-command-mode-activate)
```
It is not very important though because once you get used to it you do not want to turn the mode off:)

**Comment by xahlee** [comment]
i tried that but problem remains... see the latest code just committed 1 min ago.

**Comment by phoenixanimations** [comment]
Hello I found a possible root of the problem. In `define-minor-mode xah-fly-keys` you have 
````
(define-minor-mode xah-fly-keys
  ...
  t "∑flykeys" xah-fly-key-map
````

When I change `t "∑flykeys" xah-fly-key-map` to `nil "∑flykeys" xah-fly-key-map`. The mode turns off successfully now! It also doesn't load automatically when you do `(require 'xah-fly-keys)`. The one odd glitch is it doesn't clean up the modeline, but the hotkeys are disabled. Now I haven't looked that deep on minor modes, is there a specific reason you need it to be `t` instead of `nil`?

**Comment by xahlee** [comment]
oh, thanks!
Yes, i think i made it to t because else the mode doesn't turn on automatically after load/require.
but maybe that's the proper way to go.

**Comment by xahlee** [comment]
@phoenixanimations i tried what you suggested, but after M-x magit-status, the xfk keys don't work anymore, even when it is on.

e.g. M-x magit-status then
【space n】 should call end-of-buffer, but isn't.

**Comment by phoenixanimations** [comment]
Sorry I should of mention that I don't think that's the answer, just a hint on why it's not turning off. It is strange to me that init-value has that drastic of a change on xfk keys. But I'll look into it and get back to you.

**Comment by phoenixanimations** [comment]
Alright so I think I'm getting somewhere with this. `M-x magit-status` is a command that makes a new buffer right? If so it seems like setting the `init-value` to `nil` is causing the "intended" behavior in which the minor mode is enabled in the current buffer only. 

With setting `init-value` to nil, and adding this code: 
````
(define-globalized-minor-mode global-xah-fly-keys-mode
  xah-fly-keys xah-fly-keys)
````
And then calling `M-x global-xah-fly-keys-mode`, I am able to access [space n] from any buffer and even new ones. I can also disable the mode as well by calling global-xah-fly-keys-mode again.

Two things still bother me about this though, one is the modeline still has I/C hints. And two while that solution seems to work, I haven't fully tested it and something feels off about it.

**Comment by xahlee** [comment]
@phoenixanimations about the modeline, i can quickly answer.

if you look at the file, you'll see
(setq mode-line-front-space "C")

so, we just need to turn it off when the mode is off. (it was a quick hack.)

about fixing the mode on/off... i haven't looked into now. Will need to take me an hour to warm up, and can't do now.

in another bug report
https://github.com/xahlee/xah-fly-keys/issues/16
, you'll see the issue about swaping the whole mode/insert keys by simply changin a varuable instead of going thru all keys. That is also a major thing todo, and i think is related to this.

the bottom line is, have to look at this issue thoroughly. Since this is fundamental.

 (if i were to do it, i'd look at how viper mode is doing it. (viper mode is classic emacs emulation for vi.) and also look at how evil mode does it. And, look at whether using input method works, and see if it's proper or hack. and read the whole chapter on emacs input/keys in elisp manual.)

in summary, related to this issue is 2 things:

• turn xfk on/off properly. Need proper code, according to elisp doc. no hack.
• when switching between command/insert mode, need to make it by just change a variable's value to a different key map. Instead of setting every key in a keymap.

**Comment by DanLanglois** [comment]
re: this:

'• turn xfk on/off properly. Need proper code, according to elisp doc. no hack.
• when switching between command/insert mode, need to make it by just change a variable's value to a different key map. Instead of setting every key in a keymap.'

Taking both of these issues together, I have a pull request in for this. 

The first issue: 'When I switch the mode off with xah-fly-keys-off function it turns back on each time I use minibuffer.'

Now it removes the hook and reverts the modeline: 'the modeline still has I/C hints'.
Also, now using 'define-globalized-minor-mode'. I will note, that my impression is that global minor modes created with the define-globalized-minor-mode macro are a bit tricky. However, one point is that with a new global minor mode, that doesn't get activated in certain modes, you have a 'right place' to make exceptions. Just copy the definition 'define-globalized-minor-mode xah-fly-keys' and make the right exceptions.

Also, 'just change a variable's value to a different key map'. Well, this might not be brief, but as it turns out, controlling the active keymaps is interesting stuff.  I took a glance at this issue and first was thinking about having to create two new minor modes for each mode, add them to their hook, remove them from the minibuffer-setup-hook, etc.

However, I didn't go that route. 

I'm using the emacs variable 'emulation-mode-map-alists', which holds a list of keymap alists to use for emulation modes. It is intended for modes or packages using multiple minor-mode keymaps. Each element is a keymap alist which has the same format and meaning as minor-mode-map-alist. Of course, what we're trying to do is very simple, being able to define/change (buffer-)locally a keymap inside emulation-mode-map-alists. And, we're trying to change the my/map1 variable (locally or globally). And, it works. 
I think I got it.

One point to note is that if we are crystal clear about whether we can be setting the alist variable my/alist buffer-locally, then different buffers might have different states. I didn't think too hard about this, I only got it working to be incompatible with using buffer-local variables.
